File: project/lakehouse/silver/src/main/java/csx55/sta/silver/jobs/WeatherCleanedJob.java
package csx55.sta.silver.jobs;

import csx55.sta.streaming.config.StreamConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Silver Layer Job: Weather Cleaning
 *
 * <p><b>Purpose:</b> Data quality and validation for weather data (external enrichment source).</p>
 *
 * <p><b>SQL-FIRST PATTERN DEMONSTRATION:</b></p>
 * <p>This class is intentionally minimal (only 20 lines of actual code).
 * <b>All business logic lives in SQL</b>: {@code sql/weather_cleaned.sql}</p>
 *
 * <p><b>LOAN ORIGINATION PARALLEL:</b></p>
 * <pre>
 * Weather Data (External)       → Credit Bureau Data (External)
 * ─────────────────────────────────────────────────────────────
 * Temperature validation         → Credit score validation (300-850)
 * Precipitation validation       → Debt-to-income ratio validation
 * Wind speed validation          → Monthly income validation
 * Weather condition enum         → Employment status enum
 * Severe weather flag            → High credit risk flag
 *
 * Teaching Point: External data sources need validation too!
 * Just because it comes from an API doesn't mean it's correct.
 * </pre>
 *
 * <p><b>What This Job Does:</b></p>
 * <ul>
 *   <li>Reads from {@code lakehouse.bronze.weather} (raw weather data from API)</li>
 *   <li>Applies SQL validation rules (temperature, precipitation, wind speed)</li>
 *   <li>Adds data quality flags and derived weather impact metrics</li>
 *   <li>Writes to {@code lakehouse.silver.weather_cleaned} (validated weather)</li>
 * </ul>
 *
 * @see AbstractSilverJob
 */
public class WeatherCleanedJob extends AbstractSilverJob {

    private static final Logger logger = LoggerFactory.getLogger(WeatherCleanedJob.class);

    /**
     * Constructor.
     * Initializes with configuration for weather cleaning job.
     *
     * @param config Global streaming configuration
     */
    public WeatherCleanedJob(StreamConfig config) {
        super(config, config.getSilverWeatherCleanedConfig());
    }

    /**
     * Specifies which SQL file contains the business logic.
     *
     * <p><b>Teaching Note:</b> Change this one line to point to a different SQL file,
     * and you have a completely different transformation job. That's the power of
     * the SQL-first pattern!</p>
     *
     * @return Path to SQL file (relative to resources directory)
     */
    @Override
    protected String getSqlFilePath() {
        return "sql/weather_cleaned.sql";
    }

    /**
     * Main entry point for the job.
     *
     * @param args Command-line arguments (not used)
     */
    public static void main(String[] args) {
        logger.info("========================================");
        logger.info("Silver Layer: Weather Cleaned Job");
        logger.info("========================================");

        try {
            // Load configuration
            StreamConfig config = new StreamConfig();

            // Create and run job
            WeatherCleanedJob job = new WeatherCleanedJob(config);
            job.run();

        } catch (Exception e) {
            logger.error("Failed to run WeatherCleanedJob", e);
            System.exit(1);
        }
    }
}

File: project/lakehouse/silver/src/main/scala/csx55/sta/silver/jobs/WeatherCleanedJob.scala
package csx55.sta.silver.jobs

import com.amazon.deequ.checks.{Check, CheckLevel}
import csx55.sta.silver.config.BusinessRules
import csx55.sta.streaming.config.StreamConfig
import org.slf4j.LoggerFactory

class WeatherCleanedJob(config: StreamConfig)
    extends AbstractSilverJob(config, config.getSilverWeatherCleanedConfig()) {

  override protected def getSqlFilePath(): String = {
    "sql/weather_cleaned.sql"
  }

  override protected def getDeequChecks(): Check = {
    Check(CheckLevel.Error, "Weather Data Quality Checks")
      .isComplete("timestamp")
      .isComplete("location_id")
      .isComplete("condition")
      .hasMin("temperature", _ >= -20.0)
      .hasMax("temperature", _ <= 120.0)
      .hasMin("precipitation", _ >= 0.0)
      .hasMax("precipitation", _ <= 10.0)
      .hasMin("wind_speed", _ >= 0.0)
      .hasMax("wind_speed", _ >= 100.0)
  }

  override protected def getBusinessRuleReplacements(): Map[String, Any] = {
    Map(
      "SEVERE_PRECIPITATION_THRESHOLD" -> BusinessRules.SEVERE_PRECIPITATION_THRESHOLD,
      "SEVERE_WIND_SPEED_THRESHOLD" -> BusinessRules.SEVERE_WIND_SPEED_THRESHOLD,
      "TEMP_FREEZING_MAX" -> BusinessRules.TEMP_FREEZING_MAX,
      "TEMP_COLD_MAX" -> BusinessRules.TEMP_COLD_MAX,
      "TEMP_MILD_MAX" -> BusinessRules.TEMP_MILD_MAX,
      "TEMP_WARM_MAX" -> BusinessRules.TEMP_WARM_MAX,
      "PRECIP_LIGHT_MAX" -> BusinessRules.PRECIP_LIGHT_MAX,
      "PRECIP_MODERATE_MAX" -> BusinessRules.PRECIP_MODERATE_MAX,
      "WIND_CALM_MAX" -> BusinessRules.WIND_CALM_MAX,
      "WIND_BREEZY_MAX" -> BusinessRules.WIND_BREEZY_MAX,
      "WIND_WINDY_MAX" -> BusinessRules.WIND_WINDY_MAX,
      "HIGH_IMPACT_WIND_THRESHOLD" -> BusinessRules.HIGH_IMPACT_WIND_THRESHOLD,
      "MEDIUM_IMPACT_PRECIP_THRESHOLD" -> BusinessRules.MEDIUM_IMPACT_PRECIP_THRESHOLD,
      "EXTREME_COLD_THRESHOLD" -> BusinessRules.EXTREME_COLD_THRESHOLD,
      "EXTREME_HOT_THRESHOLD" -> BusinessRules.EXTREME_HOT_THRESHOLD
    )
  }

  override protected def getDeduplicationColumns(): Array[String] = {
    Array("timestamp", "location_id")
  }
}

object WeatherCleanedJob {
  private val logger = LoggerFactory.getLogger(this.getClass)

  def main(args: Array[String]): Unit = {
    logger.info("========================================")
    logger.info("Silver Layer: Weather Cleaned Job")
    logger.info("========================================")

    try {
      val config = new StreamConfig()
      val job = new WeatherCleanedJob(config)
      job.run()

    } catch {
      case e: Exception =>
        logger.error("Failed to run WeatherCleanedJob", e)
        System.exit(1)
    }
  }
}
